include/mfmg/dealii/lanczos.templates.hpp

```diff
@@ -0,0 +1,384 @@
+/*************************************************************************
+ * Copyright (c) 2017-2019 by the mfmg authors                           *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * This file is part of the mfmg libary. mfmg is distributed under a BSD *
+ * 3-clause license. For the licensing terms see the LICENSE file in the *
+ * top-level directory                                                   *
+ *                                                                       *
+ * SPDX-License-Identifier: BSD-3-Clause                                 *
+ *************************************************************************/
+
+#ifndef MFMG_LANCZOS_LANCZOS_TEMPLATE_HPP
+#define MFMG_LANCZOS_LANCZOS_TEMPLATE_HPP
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <random>
+#include <vector>
+
+#include "lanczos.hpp"
+#include "lanczos_deflatedop.templates.hpp"
+
+// This complex code has to be included before lapacke for the code to compile.
+// Otherwise, it conflicts with boost or Kokkos.
+#include <complex>
+#define lapack_complex_float std::complex<float>
+#define lapack_complex_double std::complex<double>
+#include <lapacke.h>
+
+namespace mfmg
+{
+
+/// \brief Lanczos solver: constructor
+template <typename OperatorType, typename VectorType>
+Lanczos<OperatorType, VectorType>::Lanczos(
+    OperatorType const &op, boost::property_tree::ptree const &params)
+    : _op(op)
+{
```

Comment:
I am not a big fan of the data member initialization in that constructor.
Some of the data members are uninitialized depending on the value of `_is_deflated` but it is not obvious when looking at the body.
I would suggest `: _op{op}, _is_deflated{params.get<bool>("is_deflated")}, ..., _num_evecs_per_cycle{_is_deflated ?params.get<int>("num_eigenpairs_per_cycle") : <some value possibly denoting it is invalid>}, ... { <constructor body with assertions, preferably no if> }`